### PR TITLE
Batch_Norm reference && tests

### PR DESCRIPTION
**Pull request by RodionovF**

### CODE REVIEW

**Comment by RodionovF** [comment]
@KSTconsult

File: tests/unit/kernels_reference.cpp [hunk 1]
```diff
@@ -153,6 +153,120 @@ void dw::reference::SGDStep(Parameters& params, float learning_rate) {
     }
 }
 
+void dw::reference::CPUBatchNorm1DForward(const float* input, float* output,
+                                          float* input_centered, float* std,
+                                          float* running_mean, float* running_var,
+                                          bool isTraining, float eps, float alpha,
+                                          const float* gamma, const float* beta,
+                                          const size_t rows, const size_t cols) {
+    if (isTraining) {
```
Comment (TolyaTalamanov) [inline tests/unit/kernels_reference.cpp:162]: not to mix `isTraining` with `running_mean`, let it be `is_training`

File: tests/unit/test_batch_norm.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
+#include <numeric>
+
+#include <deepworks/deepworks.hpp>
+
+#include "kernels_reference.hpp"
+#include "test_utils.hpp"
+
+namespace dw = deepworks;
+
+struct BatchNorm : public ::testing::Test {
```
Comment (TolyaTalamanov) [inline tests/unit/test_batch_norm.cpp:10]: BatchNormTest

File: tests/unit/kernels_reference.hpp [hunk 3]
```diff
@@ -30,6 +30,17 @@ namespace reference {
 
     void SGDStep(Parameters& params, float learning_rate);
 
+    void CPUBatchNorm1DForward(const float* input, float* output,
```
Comment (TolyaTalamanov) [inline tests/unit/kernels_reference.hpp:33]: Why don't you use `Tensor` ? I believe "old" references use raw pointer only because were developed at the moment tensor isn't finished

File: tests/unit/test_batch_norm.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        dw::initializer::zeros(gradBeta);
+        gradGamma.copyTo(ref_gradGamma);
+        gradBeta.copyTo(ref_gradBeta);
+
+        ref_gradInput = dw::Tensor{in.shape()};
+        dw::initializer::zeros(ref_gradInput);
+    }
+
+    void forward_reference(const dw::Tensor& input, dw::Tensor& output) {
+        std::vector<float> ref_moving_mean(in_features);
```
Comment (TolyaTalamanov) [inline tests/unit/test_batch_norm.cpp:34]: Why it isn't `Tensor`'s ?

File: tests/unit/kernels_reference.hpp (hunk 3 above)
Comment (RodionovF) [reply tests/unit/kernels_reference.hpp]: now i use tensor in functions signature, tests became more clear, but reference logic the same, in functions i use raw pointers. Is it ok?

Comment (TolyaTalamanov) [reply tests/unit/kernels_reference.hpp]: You can use whatever, because it will be hidden by the `Tensor`

File: tests/unit/test_batch_norm.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    dw::Tensor gamma, beta;
+
+    dw::Tensor gradGamma, gradBeta;
+    dw::Tensor ref_gradGamma, ref_gradBeta;
+
+    dw::Tensor output{model.outputs()[0].shape()};
+    dw::Tensor expected{output.shape()};
+
+    dw::Tensor input_centered{in.shape()};
+    dw::Tensor std{output.shape()};
```
Comment (KSTconsult) [inline tests/unit/test_batch_norm.cpp:72]: Std should be vector with size = in_features, not matrix with shape [batch_size, in_features]. Same with ref_moving_mean and ref_moving_var above.

File: tests/unit/test_batch_norm.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+    dw::Tensor gamma, beta;
+
+    dw::Tensor gradGamma, gradBeta;
+    dw::Tensor ref_gradGamma, ref_gradBeta;
+
+    dw::Tensor output{model.outputs()[0].shape()};
+    dw::Tensor expected{output.shape()};
+
+    dw::Tensor input_centered{in.shape()};
```
Comment (KSTconsult) [inline tests/unit/test_batch_norm.cpp:71]: For simply understanding all reference variables must have prefix ref

File: tests/unit/kernels_reference.cpp [hunk 7]
```diff
@@ -103,7 +103,7 @@ void dw::reference::CPUReLUBackward(const float* in, const float* grad_output,
                    [](float in, float go) { return in > 0.0 ? go : 0.0;});
 }
 
-float dw::reference::CPUCrossEntropyLossForward(const dw::Tensor& X, const dw::Tensor& target) {
+float dw::reference::CPUCrossEntropyLossForward(const Tensor& X, const Tensor& target) {
```
Comment (TolyaTalamanov) [inline tests/unit/kernels_reference.cpp:106]: Just wondering why did you remove `dw::`

File: tests/unit/test_batch_norm.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    // Reference
+    forward_reference(input, expected);
+    backward_reference(input, expected, grad_output);
+
+    // Assert
+    dw::testutils::AssertTensorEqual(gradGamma, ref_gradGamma);
+    dw::testutils::AssertTensorEqual(gradBeta, ref_gradBeta);
+
+    // Assert ??? do we need it? or not this
+    // dw::testutils::AssertTensorEqual(?, ref_gradInput);
```
Comment (TolyaTalamanov) [inline tests/unit/test_batch_norm.cpp:112]: Decided that not, only local check would be enough

File: tests/unit/test_batch_norm.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        gradBeta      = model.getLayer("batchnorm1d").params()[1].grad();
+        ref_gradGamma = dw::Tensor{gradGamma.shape()};
+        ref_gradBeta  = dw::Tensor{gradBeta.shape()};
+
+        dw::initializer::zeros(gradGamma);
+        dw::initializer::zeros(gradBeta);
+        gradGamma.copyTo(ref_gradGamma);
+        gradBeta.copyTo(ref_gradBeta);
+
+        ref_gradInput = dw::Tensor{in.shape()};
```
Comment (TolyaTalamanov) [inline tests/unit/test_batch_norm.cpp:29]: unused

**Review by KSTconsult — APPROVED** [review]

File: tests/unit/kernels_reference.cpp [hunk 10]
```diff
@@ -153,6 +153,147 @@ void dw::reference::SGDStep(Parameters& params, float learning_rate) {
...
+    float* raw_output         = output.data();
+    float* raw_input_centered = input_centered.data();
+    float* raw_std            = std.data();
+    float* raw_running_mean   = running_mean.data();
+    float* raw_running_var    = running_var.data();
+    float* raw_gamma          = gamma.data();
+    float* raw_beta           = beta.data();
+
+    if (is_training) {
+        std::vector<float> input_mean(in_features);
```
Comment (TolyaTalamanov) [inline tests/unit/kernels_reference.cpp:176]: So, there are a lot of loops, a bit confused why it isn't a single loop ? here and bellow as well

**Review by TolyaTalamanov — COMMENTED** [review]
BTW is there any improvements in using `BatchNorm1D` in sample ?

File: tests/unit/kernels_reference.cpp (hunk 10 above)
Comment (RodionovF) [reply tests/unit/kernels_reference.cpp:176]: removed a few extra cycles

File: samples/mnist/main.cpp [hunk 11]
```diff
@@ -69,6 +70,7 @@ int main(int argc, char *argv[]) {
         std::cout << "Loss: " << loss << std::endl;
 
         // NB: Reset val state
+        model.train(false);
```
Comment (TolyaTalamanov) [inline samples/mnist/main.cpp:73]: So, what about `model.train(true)` in training loop ? 
BTW does it work  ?

File: tests/unit/test_networks.cpp [hunk 12]
```diff
@@ -272,6 +316,28 @@ TEST_F(MNISTModel, TrainLoopSmoke) {
         }
     }
 
+    model.train(false);
```
Comment (TolyaTalamanov) [inline tests/unit/test_networks.cpp:319]: What is it ?

File: tests/unit/test_networks.cpp [hunk 13]
```diff
@@ -222,6 +276,24 @@ TEST_F(MNISTModel, Forward) {
     validate();
 }
 
+TEST_F(MNISTModel, ForwardTrainFalse) {
+    // Init
+    dw::Tensor input(in.shape());
+    dw::initializer::uniform(input);
+
+    model.train(false);
+    train = false;
```
Comment (TolyaTalamanov) [inline tests/unit/test_networks.cpp:285]: Anyway can't understand what is it ? Why it's needed `train` ? Why don't just `model.train(false)`

Comment (TolyaTalamanov) [reply tests/unit/test_networks.cpp:285]: OK, got it, the ordering is confusing
```
train = false
model.train(train)
```
or make it as a parameter not a class member
But let do it later.

**Comment by TolyaTalamanov** [comment]
Great! Thanks! :+1:
